src/args.rs

```diff
@@ -190,4 +355,74 @@ impl Args {
     ) -> IterBridge<WalkParallelIterator> {
         into_parallel_iterator(self.get_project_file_walker(), non_fatal_errors)
     }
+
+    pub fn color_specs(&self) -> ColorSpecs {
+        let mut specs = default_color_specs();
+        for user_color_spec in &self.colors {
+            specs.push(user_color_spec.clone());
+        }
+        ColorSpecs::new(&specs)
+    }
+
+    pub fn buffer_writer(&self) -> BufferWriter {
+        let mut wtr = BufferWriter::stdout(self.color_choice());
+        wtr.separator(self.file_separator());
+        wtr
+    }
+
+    fn color_choice(&self) -> ColorChoice {
+        match self.color.unwrap_or(ColorChoiceArg::Auto) {
+            ColorChoiceArg::Always => ColorChoice::Always,
+            ColorChoiceArg::Ansi => ColorChoice::AlwaysAnsi,
+            ColorChoiceArg::Auto => {
+                if grep_cli::is_tty_stdout() || self.pretty {
+                    ColorChoice::Auto
+                } else {
+                    ColorChoice::Never
+                }
+            }
+            ColorChoiceArg::Never => ColorChoice::Never,
+        }
+    }
+
+    fn heading(&self) -> bool {
+        if self.no_heading || self.vimgrep {
+            false
+        } else {
+            grep_cli::is_tty_stdout() || self.heading || self.pretty
+        }
+    }
+
+    fn file_separator(&self) -> Option<Vec<u8>> {
+        // if self.output_kind() != OutputKind::Standard {
+        //     return Ok(None);
+        // }
+
+        let (ctx_before, ctx_after) = self.contexts();
+        if self.heading() {
+            Some(b"".to_vec())
+        } else if ctx_before > 0 || ctx_after > 0 {
+            self.context_separator()
```

Comment:
This is what is responsible for the fact that now in some of the existing test diffs you'll see that there are `--` lines between files when passing a "context" option

(so we weren't matching `ripgrep`'s behavior there basically, now we are)